File: backend/goals.py
import os
import uuid
from datetime import datetime, timezone
from typing import Optional

import boto3
from boto3.dynamodb.conditions import Key
from fastapi import APIRouter, HTTPException
from pydantic import BaseModel

router = APIRouter(prefix="/goals", tags=["goals"])

GOALS_TABLE = os.environ.get("GOALS_TABLE", "puran-goals-dev")
USER_ID = os.environ.get("USER_ID", "demo-user")

_dynamodb = boto3.resource("dynamodb")
table = _dynamodb.Table(GOALS_TABLE)


class GoalCreate(BaseModel):
    title: str
    targetAmount: float
    deadline: str
    type: str  # "short" | "veryShort"


class GoalUpdate(BaseModel):
    currentAmount: Optional[float] = None
    status: Optional[str] = None  # pending | achieved | failed
    title: Optional[str] = None
    targetAmount: Optional[float] = None
    deadline: Optional[str] = None


@router.get("")
def list_goals():
    resp = table.query(KeyConditionExpression=Key("userId").eq(USER_ID))
    return {"goals": resp.get("Items", [])}


@router.post("", status_code=201)
def create_goal(goal: GoalCreate):
    item = {
        "userId": USER_ID,
        "goalId": str(uuid.uuid4()),
        "title": goal.title,
        "targetAmount": str(goal.targetAmount),
        "currentAmount": "0",
        "deadline": goal.deadline,
        "type": goal.type,
        "status": "pending",
        "createdAt": datetime.now(timezone.utc).isoformat(),
    }
    table.put_item(Item=item)
    return item


@router.put("/{goal_id}")
def update_goal(goal_id: str, update: GoalUpdate):
    field_map = {
        "currentAmount": ("#ca", ":ca", str),
        "status":        ("#st", ":st", str),
        "title":         ("#ti", ":ti", str),
        "targetAmount":  ("#ta", ":ta", str),
        "deadline":      ("#dl", ":dl", str),
    }
    expr_parts, names, values = [], {}, {}
    for field, (name_key, val_key, transform) in field_map.items():
        val = getattr(update, field)
        if val is not None:
            expr_parts.append(f"{name_key} = {val_key}")
            names[name_key] = field
            values[val_key] = transform(val)

    if not expr_parts:
        raise HTTPException(status_code=400, detail="No fields to update")

    resp = table.update_item(
        Key={"userId": USER_ID, "goalId": goal_id},
        UpdateExpression="SET " + ", ".join(expr_parts),
        ExpressionAttributeNames=names,
        ExpressionAttributeValues=values,
        ReturnValues="ALL_NEW",
    )
    return resp.get("Attributes", {})


@router.delete("/{goal_id}", status_code=204)
def delete_goal(goal_id: str):
    table.delete_item(Key={"userId": USER_ID, "goalId": goal_id})

File: frontend/src/api/goals.js
import { api } from "./client";

const DATASET_ID = import.meta.env.VITE_DATASET_ID || import.meta.env.VITE_USER_ID || "demo";

function normalizeGoal(goal) {
  if (!goal) {
    return goal;
  }

  const targetAmount = goal.targetAmount ?? goal.target_amount ?? goal.TotalAmount ?? 0;
  const currentAmount = goal.currentAmount ?? goal.amount_saved ?? goal.AmountSaved ?? 0;
  const deadline = goal.deadline ?? goal.end_date ?? goal.EndDate ?? "";
  const title = goal.title ?? goal.description ?? goal.Description ?? "Untitled goal";
  const type = goal.type ?? goal.category ?? goal.Category ?? "short";
  const result = goal.result ?? goal.Result;

  return {
    ...goal,
    goalId: goal.goalId ?? goal.goal_id ?? goal.SK?.replace("GOAL#", ""),
    title,
    targetAmount: Number(targetAmount),
    currentAmount: Number(currentAmount),
    deadline,
    type,
    status: goal.status ?? (result ? "achieved" : "pending"),
  };
}

function normalizeGoals(payload) {
  const goals = Array.isArray(payload) ? payload : payload?.goals || payload?.items || [];

  return goals.map(normalizeGoal);
}

export async function getGoals() {
  return normalizeGoals(await api("/goals"));
}

export async function createGoal(goal) {
  return normalizeGoal(await api("/goals", {
    method: "POST",
    body: JSON.stringify(goal),
  }));
}

export async function updateGoal(goalId, updates) {
  return api(`/goals/${goalId}`, {
    method: "PUT",
    body: JSON.stringify({
      dataset_id: updates.dataset_id || DATASET_ID,
      ...updates,
    }),
  });
}

export async function deleteGoal(goalId) {
  return api(`/goals/${goalId}`, {
    method: "DELETE",
  });
}
